rust/lance/src/io/stream.rs

```diff
@@ -12,14 +12,42 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
 use arrow_array::RecordBatch;
 use arrow_schema::SchemaRef;
-use futures::Stream;
+use futures::{stream::BoxStream, Stream, StreamExt};
 
 use crate::Result;
 
 /// RecordBatch Stream trait.
-pub trait RecordBatchStream: Stream<Item = Result<RecordBatch>> + Send + 'static {
+pub trait RecordBatchStream<'a>: Stream<Item = Result<RecordBatch>> + Send + 'a {
     /// Returns the schema of the stream.
     fn schema(&self) -> SchemaRef;
 }
+
+pub struct BoxedRecordBatchStream<'a> {
+    inner: BoxStream<'a, Result<RecordBatch>>,
+    schema: SchemaRef,
+}
+
+impl<'a> BoxedRecordBatchStream<'a> {
```

Comment:
You should check out DataFusion's [RecordBatchStreamAdapter](https://docs.rs/datafusion/31.0.0/src/datafusion/physical_plan/stream.rs.html#249-297). It does the same thing, but is generic over the stream, so can also handle unboxed cases.

One thing it does that's not here is supplying `size_hint()`, so I'd at least add that.

Reply:
Ok, I can implement that, similar to our `RecordBatchStream` which returns our `Result<Batch, lance::Error>`